server/src/main/java/org/apache/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -165,59 +168,141 @@ public int getBufferSize()
     private final ConcurrentHashMap<String, Long> producerSequences = new ConcurrentHashMap<>();
-    private final Stopwatch idleWatch = Stopwatch.createUnstarted();
 
-    public EventReceiverFirehose(InputRowParser<Map<String, Object>> parser)
+    /**
+     * This field and {@link #requestedShutdownTimeNsHolder} use nanoseconds instead of milliseconds not to deal with
+     * the fact that {@link System#currentTimeMillis()} can "go backward", e. g. due to time correction on the server.
+     */
+    private final AtomicReference<Long> idleCloseTimeNsHolder = new AtomicReference<>();
+    private final AtomicReference<Long> requestedShutdownTimeNsHolder = new AtomicReference<>();
+
+    EventReceiverFirehose(InputRowParser<Map<String, Object>> parser)
     {
       this.buffer = new ArrayBlockingQueue<>(bufferSize);
       this.parser = parser;
-      exec = Execs.scheduledSingleThreaded("event-receiver-firehose-%d");
-      idleDetector = Execs.singleThreaded("event-receiver-firehose-idle-detector-%d");
-      idleDetector.submit(() -> {
-        long idled;
-        try {
-          while ((idled = idleWatch.elapsed(TimeUnit.MILLISECONDS)) < maxIdleTime) {
-            Thread.sleep(maxIdleTime - idled);
-          }
-        }
-        catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          return;
+
+      if (maxIdleTimeMillis != Long.MAX_VALUE) {
+        idleCloseTimeNsHolder.set(System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(maxIdleTimeMillis));
+        synchronized (this) {
+          createDelayedCloseExecutor();
         }
-        log.info("Firehose has been idle for %d ms, closing.", idled);
-        close();
-      });
-      idleWatch.start();
+      }
     }
 
+    @GuardedBy("this")
+    private Thread createDelayedCloseExecutor()
+    {
+      Thread delayedCloseExecutor = new Thread(
+          () -> {
+            // The closed = true is visible after close() because there is a happens-before edge between
+            // delayedCloseExecutor.interrupt() call in close() and catching InterruptedException below in this loop.
+            while (!closed) {
+              Long closeTimeNs = null;
+              Boolean dueToShutdownRequest = null;
+              Long idleCloseTimeNs = idleCloseTimeNsHolder.get();
+              if (idleCloseTimeNs != null) {
+                closeTimeNs = idleCloseTimeNs;
+                dueToShutdownRequest = false;
+              }
+              Long requestedShutdownTimeNs = requestedShutdownTimeNsHolder.get();
+              if (requestedShutdownTimeNs != null) {
+                if (closeTimeNs == null || requestedShutdownTimeNs - closeTimeNs <= 0) { // overflow-aware comparison
+                  closeTimeNs = requestedShutdownTimeNs;
+                  dueToShutdownRequest = true;
+                }
+              }
+              if (closeTimeNs == null) {
+                log.error(
```

Comment:
Could we come up with a unit-test which can detect this error?

Reply:
Added a comment to code. Although it kind of duplicates the error message in the logging statement, admittedly text inside logging statements is a "blind spot", we don't read it when reading code.

Reply:
I wonder if a unit-test possible which can spot such a bug in code?

Reply:
Maybe, but the benefit of such a unit test doesn't justify the effort of creating it for me.

Reply:
Could you write a sanity test for the method which checks the class correctly works with cases 
```
idleCloseTimeNs  < requestedShutdownTimeNs
idleCloseTimeNs  == requestedShutdownTimeNs
idleCloseTimeNs > requestedShutdownTimeNs
```

Reply:
I changed some existing tests to check that that `delayedCloseExecutor` exits.

Reply:
Just wonder why not more concise
```java
while (!closed) {
  if (requestedShutdownTimeNs !=null && requestedShutdownTimeNs - System.nanoTime() <= 0) {
    log.info(...)
    close()
  }
  else if (idleCloseTimeNs - System.nanoTime() <= 0) {
    log.info(...)
    close()
  }

  Threads.sleepForWithCatchInside(closeTimeoutNs, TimeUnit.NANOSECONDS);
}
```

Reply:
Might be NPE in `idleCloseTimeNs`. Is this issue a blocker for this PR?

Reply:
Couldn't we initialize idleCloseTimeNs = Long.MAX_VALUE at startup?

Reply:
`null` in `idleCloseTimeNs` is used as an indication that we don't track idle time. (It's possible to check that separately in the delayedCloseExecutor, but that's extra complexity too.)

BTW `Long.MAX_VALUE` is not special, `System.nanoTime()` may easily be `Long.MAX_VALUE - 1`.

Reply:
```java
while (!closed) {
  if (requestedShutdownTimeNs !=null && requestedShutdownTimeNs - System.nanoTime() <= 0) {
    log.info(...)
    close()
  }
  else if (idleCloseTimeNs != null && idleCloseTimeNs - System.nanoTime() <= 0) {
    log.info(...)
    close()
  }

  Threads.sleepForWithCatchInside(closeTimeoutNs, TimeUnit.NANOSECONDS);
}
```
We can just add this check to `else if` block. That's still more shorter that way.

Reply:
It won't allow discovering the bug in the code.